bluesky/plans.py

```diff
@@ -57,59 +57,18 @@ def count(detectors, num=1, delay=None, *, md=None):
     _md.update(md or {})
     _md['hints'].setdefault('dimensions', [(('time',), 'primary')])
 
-    # If delay is a scalar, repeat it forever. If it is an iterable, leave it.
-    if not isinstance(delay, Iterable):
-        delay = itertools.repeat(delay)
-    else:
-        try:
-            num_delays = len(delay)
-        except TypeError:
-            # No way to tell in advance if we have enough delays.
-            pass
-        else:
-            if num - 1 > num_delays:
-                raise ValueError("num=%r but delays only provides %r "
-                                 "entries" % (num, num_delays))
-        delay = iter(delay)
-
-    @bpp.stage_decorator(detectors)
-    @bpp.run_decorator(md=_md)
-    def finite_plan():
-        for i in range(num):
-            now = time.time()  # Intercept the flow in its earliest moment.
+    def outer():
+        def inner():
             yield Msg('checkpoint')
             yield from bps.trigger_and_read(detectors)
-            try:
-                d = next(delay)
-            except StopIteration:
-                if i + 1 == num:
-                    break
-                else:
-                    # num specifies a number of iterations less than delay
-                    raise ValueError("num=%r but delays only provides %r "
-                                     "entries" % (num, i))
-            if d is not None:
-                d = d - (time.time() - now)
-                if d > 0:  # Sleep if and only if time is left to do it.
-                    yield Msg('sleep', None, d)
+        return (yield from inner())
 
     @bpp.stage_decorator(detectors)
     @bpp.run_decorator(md=_md)
-    def infinite_plan():
-        while True:
-            yield Msg('checkpoint')
-            yield from bps.trigger_and_read(detectors)
-            try:
-                d = next(delay)
-            except StopIteration:
-                break
-            if d is not None:
-                yield Msg('sleep', None, d)
+    def plan():
```

Comment:
While we're changing this, we should follow a pattern established @klauer some time after `count` was written: naming interior generator functions `inner_*` -- i.e. this would be called `inner_count`. If all goes well, users never see these names, but they can make tracebacks more readable. See for example `scan_nd` and its local variable `inner_scan_nd`, etc.